### PR TITLE
[triage] Allow the triaging of a test, unrelated to a product

### PR DESCRIPTION
**Pull request by stephenmcgruer**
This adds support for triaging a test unrelated to a product (e.g.
Chrome). The goal is to allow the triaging of tests that are problematic
due to the test itself (e.g. if the test violates the spec or has bugs),
rather than problematic due to a browser bug.

To achieve this without changing the current data model, we abuse the
product notation and add a new product named 'test-issue', for issues
with tests. Metadata for this product are shown next to the test name
rather than in the results boxes, and similarly one triages test issues
by selecting the test name cell.

### CODE REVIEW

**Comment by stephenmcgruer** [comment]
@KyleJu PTAL. There are a few triaged test-level cases in `/css/css-align/parsing` for you to look at, and https://github.com/web-platform-tests/wpt-metadata/pull/920 and https://github.com/web-platform-tests/wpt-metadata/pull/919 were PRs made by this change. (I'll clean them up after landing).

I suspect that searching won't quite work as expected here, but I hope we can fix that as a follow-up.

**Comment by stephenmcgruer** [comment]
@KyleJu ping for review? The rendering team requested this fairly urgently, and there are other requests that need to build on top of this, so this is blocking work.

File: webapp/views/wpt-results.js [hunk 1]
```diff
@@ -226,8 +226,13 @@ class WPTResults extends AmendMetadataMixin(Pluralizer(WPTColors(WPTFlags(PathIn
                     </template>
                   </template>
                 </td>
-                <td>
+                <td onclick="[[handleTriageSelect(null, node, testRun)]]" onmouseover="[[handleTriageHover(null, node, testRun)]]">
```
Comment (KyleJu) [inline webapp/views/wpt-results.js:229]: I don't think this is quite working on the UI. I can see ` this.handleHover` being triggered when I hover over, but I don't see any css highlights nor will you be able to triage that cell

File: webapp/views/wpt-results.js [hunk 2]
```diff
@@ -226,8 +226,13 @@ class WPTResults extends AmendMetadataMixin(Pluralizer(WPTColors(WPTFlags(PathIn
                     </template>
                   </template>
                 </td>
-                <td>
```
Comment (KyleJu) [inline webapp/views/wpt-results.js:229]: Does this feature work on subtests? if so we have to add the same code to [`test-files-results-table`](https://github.com/web-platform-tests/wpt.fyi/blob/main/webapp/components/test-file-results-table.js#L140)

Comment (stephenmcgruer) [reply webapp/views/wpt-results.js:229]: It does not.

File: webapp/components/wpt-amend-metadata.js [hunk 3]
```diff
@@ -252,13 +252,26 @@ class AmendMetadata extends LoadingState(PathInfo(ProductInfo(PolymerElement)))
...
+            metadata['product'] = entry.product;
+          }
+          link[test].push(metadata);
         }
       }
     }
     return link;
   }
 
+  getProductLogo(product) {
```
Comment (KyleJu) [inline webapp/components/wpt-amend-metadata.js:266]: Screen Shot 2021-02-23 at 12 37 23 PM

The icons aren't showing for the wpt-metadata entries on the bottom either. I think we need to change `this.displayLogo(product)` instead; [see wpt-metadata.js](https://github.com/web-platform-tests/wpt.fyi/blob/main/webapp/components/wpt-metadata.js#L45)

Comment (KyleJu) [reply webapp/components/wpt-amend-metadata.js:266]: Remove this method and put the same code to `displayLogo`

**Comment by KyleJu** [comment]
> @KyleJu PTAL. There are a few triaged test-level cases in `/css/css-align/parsing` for you to look at, and [web-platform-tests/wpt-metadata#920](https://github.com/web-platform-tests/wpt-metadata/pull/920) and [web-platform-tests/wpt-metadata#919](https://github.com/web-platform-tests/wpt-metadata/pull/919) were PRs made by this change. (I'll clean them up after landing).
> 
> I suspect that searching won't quite work as expected here, but I hope we can fix that as a follow-up.

Right, to fix searchcache, we can simply ignore empty products in [`PrepareLinkFilter`](https://github.com/web-platform-tests/wpt.fyi/blob/f8d9beebe98a4d254d1ae0d0283c4fe7cef3196e/shared/metadata.go#L153). This function prepares the metadata for [filter methods](https://github.com/web-platform-tests/wpt.fyi/blob/f8d9beebe98a4d254d1ae0d0283c4fe7cef3196e/api/query/cache/index/filter.go#L217) in searchcache

File: webapp/views/wpt-results.js [hunk 4]
```diff
@@ -871,6 +876,11 @@ class WPTResults extends AmendMetadataMixin(Pluralizer(WPTColors(WPTFlags(PathIn
   }
 
   canAmendMetadata(node, index, testRun) {
+    // It is always possible in triage mode to amend metadata for a problem
+    // with a test file itself.
+    if (index == null)
```
Comment (KyleJu) [inline webapp/views/wpt-results.js:881]: Lets add '{}' around this `if (index == null)`

File: webapp/views/wpt-results.js [hunk 5]
```diff
@@ -1044,7 +1054,8 @@ class WPTResults extends AmendMetadataMixin(Pluralizer(WPTColors(WPTFlags(PathIn
         return;
       }
 
-      this.handleSelect(e.target.closest('td'), this.displayedProducts[index].browser_name, node.path, this.$['selected-toast']);
+      const product = index == null ? '' : this.displayedProducts[index].browser_name;
```
Comment (KyleJu) [inline webapp/views/wpt-results.js:1057]:
```suggestion
      const product = index === null ? '' : this.displayedProducts[index].browser_name;
```

File: webapp/views/wpt-results.js (hunk 4 above)
Comment (KyleJu) [inline webapp/views/wpt-results.js:881]:
```suggestion
    if (index === null)
```

File: webapp/views/wpt-results.js [hunk 6]
```diff
@@ -1065,13 +1076,14 @@ class WPTResults extends AmendMetadataMixin(Pluralizer(WPTColors(WPTFlags(PathIn
       testname = testname + '/*';
     }
 
-    const key = testname + this.displayedProducts[index].browser_name;
+    const browserName = index == null ? '' : this.displayedProducts[index].browser_name;
```
Comment (KyleJu) [inline webapp/views/wpt-results.js:1079]:
```suggestion
    const browserName = index === null ? '' : this.displayedProducts[index].browser_name;
```

**Review by KyleJu — APPROVED** [review]
LGMT % nits

File: webapp/views/wpt-results.js (hunk 4 above)
Comment (stephenmcgruer) [reply webapp/views/wpt-results.js]: Done.

Comment (stephenmcgruer) [reply webapp/views/wpt-results.js]: Done.

File: webapp/views/wpt-results.js (hunk 5 above)
Comment (stephenmcgruer) [reply webapp/views/wpt-results.js]: Done.

File: webapp/views/wpt-results.js (hunk 6 above)
Comment (stephenmcgruer) [reply webapp/views/wpt-results.js]: Done.

**Comment by stephenmcgruer** [comment]
Fun, the triple-equals on `null` has revealed that we're getting `undefined` for index at some point instead of `null`... need to investigate that now :(

**Pull request merged**